Add LiteLLM-based Agent with Automatic Context Trimming

NISH1001 opened the pull request:
### Summary :memo:

This pull request introduces a new **`LiteLLMInstructorBaseAgent`**, which integrates `LiteLLM` to provide broader LLM provider support and robust context management. It directly inherits from `akd.agents._base.InstructorBaseAgent` with only required changes for trimming and `acompleition` engine.

The key feature of this new agent is **automatic message trimming**, designed to prevent token limit errors by intelligently managing the conversation history size before making an API call.

Additionally, this PR includes a comprehensive **refactor of the entire test suite** for all base agents. The previous monolithic test file has been replaced with a structured, modular directory, significantly improving test clarity, organization, and maintainability.

### Details

*Describe more what you did on changes.*

1.  **Added `LiteLLMInstructorBaseAgent`**:

      * Created a new agent class in `akd/agents/_base.py` that inherits from `InstructorBaseAgent`.
      * It uses `instructor.from_litellm(acompletion)` to allow compatibility with any model supported by LiteLLM (e.g., local Ollama models, Cohere, Anthropic, etc.).
      * It automatically trims the message history using `litellm.utils.trim_messages` if `enable_trimming` is `True` in the configuration.

2.  **Enhanced `BaseAgentConfig`**:

      * Added three new fields to `BaseAgentConfig` to control the new context trimming feature:
          * `max_tokens`: Sets the context window size for the model.
          * `trim_ratio`: Defines the target percentage of `max_tokens` to aim for after trimming.
          * `enable_trimming`: A boolean to toggle the feature on or off.

3.  **Refactored Agent Tests**:

      * Deleted the old, large `tests/agents/test_base_agents.py` file.
      * Created a new test package at `tests/agents/base/`.
      * Split tests into logical files: `test_config.py`, `test_lang_base.py`, `test_instructor_base.py`, `test_litellm_base.py`, and `test_shared.py`.
      * A new `conftest.py` was added to centralize test schemas, agent implementations, and mock fixtures, reducing code duplication.

### Bugfixes :bug:

  - Fixed an issue where the `base_url` was not being correctly passed to the LiteLLM client. The code now ensures the URL is properly formatted using `str(self.base_url).rstrip("/")`, which is crucial for connecting to local model servers like Ollama.

### Usage

Here is an example of how to use the new `LiteLLMInstructorBaseAgent` to connect to a local Ollama model.

```python
from akd.agents._base import BaseAgentConfig
# Assuming QueryAgentWithLiteLLM is your implementation of LiteLLMInstructorBaseAgent
# from your_agents import QueryAgentWithLiteLLM 

# Configure the agent to use a local Ollama model
config = BaseAgentConfig(
    model_name="ollama/gemma:3b",  # LiteLLM model string
    base_url="http://localhost:11434",
    stateless=True,
    enable_trimming=True, # Enable automatic context trimming
    debug=True,
)

# Instantiate the agent
agent = QueryAgentWithLiteLLM(config)

# Run the agent
# result = await agent.arun({"query": "What is the mission of NASA IMPACT?"})
# print(result)
```

### Checks

  - [x] Tested Changes (346 passed, 78% coverage)
  - [x] Stakeholder Approval

---

## Notes

- This PR also directly addresses long-standing feature request to add platform agnostic unified model calling as documented here: https://github.com/NASA-IMPACT/accelerated-discovery/issues/74
- Additionally, this is non-destructive. There is technically no change in implementation of any downstream agents. We need to strategically move the agent to new base which is to be another followup PR.

NISH1001 committed:
Add lite-llmbased base agent

- Add `LiteLLMInstructorBaseAgent` that has trimming functionality

NISH1001 committed:
Add tests for litellmbased base agent

NISH1001 committed:
Refactor tests for base agents

- Add `tests/agents/base/`

NISH1001 committed:
Bugfix litellminstructorbaseagent with base url

- Pass `base_url` if provided

NISH1001 commented on akd/agents/_base.py:
```diff
@@ -28,6 +30,20 @@ class BaseAgentConfig(BaseConfig):
         description="Whether to maintain conversation history/state",
     )
 
+    # Token management
+    max_tokens: int = Field(
+        default=100000,
+        description="Maximum tokens for message context",
+    )
+    trim_ratio: float = Field(
+        default=0.75,
+        description="Target ratio after trimming (0.75 = use 75% of max)",
+    )
```
These are also defaults in litellm trim_messages(...)

muthukumaranR replied:
Should this be auto generated based on the model config?

muthukumaranR replied:
Also a nitpick to add Field constraints (e.g., trim_ratio in [0,1], max_tokens < model_specific.max_token, or a known mini-max number)

muthukumaranR replied:
Hopefully LiteLLM handles this though - haven't tested this myself

NISH1001 replied:
Ya probably model-based we can do. Let me quickl ylook at that. The default behaviour litellm has are these default values apparently.

NISH1001 replied:
Okay i have added the validation. litellm already has `get_model_info` function. So, i built the validator around that logic.

NISH1001 reviewed (commented):
Comment I

NISH1001 committed:
Add field constraints to trim_ratio

- Constraint `trim_ratio` to (0, 1]

NISH1001 committed:
Add token size validation in base agent config

- Add model validator to `max_tokens`
- Refactored tests to only use gpt-4o-mini in the tests
